Refactor RunApp and FileTester

milljm opened the pull request:
In order to eliminate duplicate output, have all testers which work with files
use FileTester instead of RunApp. Move several file testing methods out of RunApp
which was being used by only a couple testers, into FileTester.

Adding a unittest for ImageDiff tester. Even though unittesting occurs in Chigger,
I felt having a simple test in the TestHarness wise.

Actually run arbitrary commands.

Closes #18974

Commit:
Refactor RunApp and FileTester

In order to eliminate duplicate output, have all testers which work with files
use FileTester instead of RunApp. Move several file testing methods out of RunApp
which was being used by only a couple testers, into FileTester.

Adding a unittest for ImageDiff tester. Even though unittesting occurs in Chigger,
I felt having a simple test in the TestHarness wise.

Actually run arbitrary commands.

Closes #18974

Commit:
Incluse PATH to test_dir

When using the arbitrary command function, prefix PATH to test_spec dir.

Commit:
Include spec cli_commands

Allow cli arguments to be passed to arbitrary command.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/839940/) on 5b4ae02 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/18988/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Generate and verify coverage](https://civet.inl.gov/job/839952/) on 5b4ae02 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>ce01b1</th>
      <th colspan=4>#18988 5b4ae0</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>81.61%</td>
      <td>81.61%</td>
      <td>-0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>69703</td>
      <td>69702</td>
      <td>-1</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15702</td>
      <td>15703</td>
      <td>+1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/18988/coverage_diff/framework.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/18988/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/18988/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/18988/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/18988/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/18988/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/18988/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/18988/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/18988/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/18988/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/18988/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/18988/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/18988/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/18988/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/18988/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/18988/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/18988/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/18988/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/18988/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/18988/coverage/rdg)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/18988/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/18988/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/18988/coverage/tensor_mechanics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/18988/coverage/xfem)

</details>



This comment will be updated on new commits.

Commit:
Remove the not-used imagediff test spec

I forgot to remove this would-have-been unittest test spec file.
There is still a unittest being done using an existing test spec file.
